src/lib.rs

```diff
@@ -391,24 +402,24 @@ pub fn compile_module<'c, T: SourceContainer>(
 
     // ### PHASE 1 ###
     // parse & index everything
-    for container in sources {
-        let location: String = container.get_location().into();
-        let e = container
-            .load_source(encoding)
-            .map_err(|err| Diagnostic::io_read_error(location.as_str(), err.as_str()))?;
-
-        let (mut parse_result, diagnostics) =
-            parser::parse(lexer::lex_with_ids(e.source.as_str(), id_provider.clone()));
-
-        //pre-process the ast (create inlined types)
-        ast::pre_process(&mut parse_result, id_provider.clone());
-        //index the pou
-        full_index.import(index::visitor::visit(&parse_result, id_provider.clone()));
-
-        //register the file with the diagnstician, so diagnostics are later able to show snippets from the code
-        let file_id = diagnostician.register_file(location.clone(), e.source);
-        all_units.push((file_id, diagnostics, parse_result));
-    }
+    parse_and_index(
```

Comment:
I think returning the units vector and the index here in the result instead of passing them as parameters, and then just adding all to the all_units / merging the index would be easier to read, and later easier to split/run in parallel.